test/test_storage_record.py

```diff
@@ -61,6 +62,61 @@ def test_mandatory_fields(self):
         except Exception, e:
             self.fail('_check_fields method failed: %s [%s]' % (str(e), str(type(e))))
 
+    def test_get_apel_db_insert(self):
+        """Check that calling get_apel_db_insert raises no exceptions
+
+        Previously there was a mismatch between the number of arguments of this
+        and the parent Record class which led to a TypeError being raised.
+        """
+
+        record = StorageRecord()
+        record.set_field('RecordId', 'host.example.org/sr/87912469269276')
+        record.set_field('CreateTime', 1289293612)
+        record.set_field('StorageSystem', 'host.example.org')
+        record.set_field('StartTime', 1286785900)
+        record.set_field('EndTime', 1286786500)
+        record.set_field('ResourceCapacityUsed', 14728)
+
+        record.get_apel_db_insert()
+        record.get_apel_db_insert(None)
+
+    def test_get_ur(self):
+        """Check XML output of get_ur
+
+        This checks that all input values save for createTime are present in the
+        XML output, then it parses the XML back into a storage record and checks
+        that the only mismatching field is CreateTime.
+        """
+        data = ('host.example.org/sr/87912469269276', 1289293612,
+                'host.example.org', 'MySite', 'pool-003', 'disk', 'replicated',
+                42, '/home/projectA', 'johndoe', 'projectA',
+                '/O=Grid/OU=example.org/CN=John Doe',
+                'binarydataproject.example.org', 'uk', 'poweruser',
+                '2010-10-11T09:31:40Z', '2010-10-11T09:41:40Z', 14728, 13617, 0)
+
+        record = StorageRecord()
+        record.load_from_tuple(data)
+        ur = record.get_ur()
+
+        for value in data:
+            # Skip the createTime value as it's regenerated when calling get_ur
+            if value != 1289293612 and str(value) not in ur:
+                self.fail("Input value '%s' not found in XML output." % value)
+
+        XML_HEADER = '<?xml version="1.0" ?>'
+        UR_OPEN = ('<sr:StorageUsageRecords xmlns:sr="http://eu-emi.eu/namespac'
+                   'es/2011/02/storagerecord">')
+        UR_CLOSE = '</sr:StorageUsageRecords>'
+
+        parser = StarParser(XML_HEADER + UR_OPEN + ur + UR_CLOSE)
+
+        # This works as all dictionary values are hashable/immutable
+        missmatched = (set(parser.get_records()[0]._record_content.items()) ^
```

Comment:
what does `^` do here? If I read it correctly its an operation on two `set` objects?

Reply:
It's the "symmetric difference". It creates a new set with elements in either the first or second set but not both: https://docs.python.org/2/library/sets.html#set-objects

Reply:
ah okay, shiny!